core/src/main/java/io/questdb/griffin/engine/table/LatestByAllSymbolsFilteredRecordCursor.java

```diff
@@ -0,0 +1,130 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.table;
+
+import io.questdb.cairo.RecordSink;
+import io.questdb.cairo.map.Map;
+import io.questdb.cairo.map.MapKey;
+import io.questdb.cairo.sql.DataFrame;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.SqlExecutionCircuitBreaker;
+import io.questdb.cairo.sql.StaticSymbolTable;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.constants.BooleanConstant;
+import io.questdb.std.DirectLongList;
+import io.questdb.std.IntList;
+import io.questdb.std.Misc;
+import io.questdb.std.Rows;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+class LatestByAllSymbolsFilteredRecordCursor extends AbstractDescendingRecordListCursor {
+
+    private static final Function NO_OP_FILTER = BooleanConstant.TRUE;
+    private final Function filter;
+    private final Map map;
+    private final IntList partitionByColumnIndexes;
+    private final IntList partitionBySymbolCounts;
+    private final RecordSink recordSink;
+
+    public LatestByAllSymbolsFilteredRecordCursor(
+            @NotNull Map map,
+            @NotNull DirectLongList rows,
+            @NotNull RecordSink recordSink,
+            @Nullable Function filter,
+            @NotNull IntList columnIndexes,
+            @NotNull IntList partitionByColumnIndexes,
+            @Nullable IntList partitionBySymbolCounts
+    ) {
+        super(rows, columnIndexes);
+        this.map = map;
+        this.recordSink = recordSink;
+        this.filter = filter != null ? filter : NO_OP_FILTER;
+        this.partitionByColumnIndexes = partitionByColumnIndexes;
+        this.partitionBySymbolCounts = partitionBySymbolCounts;
+    }
+
+    @Override
+    public void close() {
+        if (isOpen()) {
+            Misc.free(filter);
+            Misc.free(map);
+            super.close();
+        }
+    }
+
+    private long countSymbolCombinations() {
+        long combinations = 1;
+        for (int i = 0, n = partitionByColumnIndexes.size(); i < n; i++) {
+            int symbolCount = partitionBySymbolCounts != null ? partitionBySymbolCounts.getQuick(i) : Integer.MAX_VALUE;
+            assert symbolCount > 0;
+            int columnIndex = partitionByColumnIndexes.getQuick(i);
+            StaticSymbolTable symbolMapReader = dataFrameCursor.getSymbolTable(columnIndexes.getQuick(columnIndex));
+            int distinctSymbols = symbolMapReader.getSymbolCount();
+            if (symbolMapReader.containsNullValue()) {
+                distinctSymbols++;
+            }
```

Comment:
Overflow is possible here if e.g. there's 6 symbol columns with 1k+ symbols each .

Reply:
I considered the overflow, but decided not to bother with it since it's not a problem when we have a negative value. But multiple overflows are indeed problematic. Fixed the bug in ff3fabdde2e4db861bb2c4f6a61510ce48e92362